### DISCUSSION TITLE
Acceptable approach for drools-lsp?

### DISCUSSION THREAD

**Discussion by errantepiphany** [mason-org/mason.nvim#924] [Q&A]
[[ Eeek, I just realized I submitted this Q&A to mason.nvim instead of mason-lspconfig.nvim. Sorry! ]]

Hi @williamboman, I'm posting this here because I would prefer not submitting a PR (to the mason-lspconfig.nvim repo and to neovim/lspconfig.nvim) until I've gotten a thumbs-up from you on its viability.

You had mentioned here: https://github.com/williamboman/mason-lspconfig.nvim/pull/128#discussion_r1071339955 that it's probably not a good idea to put the client init information in a `settings={}` block, since that gets sent to the server (which doesn't need it). So, I want to change that.

Also, I wanted to add something to mason-lspconfig.nvim so that if a person is using mason to install drools-lsp, they can do so with zero-configuration, using the `server_configurations` overrides mechanism you showed me (also please recall how we had discussed how the order-of-operations with the current code had made that difficult). So I have a proposal, code included, to get your thoughts.

First, to show how short and sweet it is on the mason-lspconfig.nvim side: https://github.com/errantepiphany/mason-lspconfig.nvim/blob/13fe2ff000dee07ab39cb665b9b2afa8db6bd5ab/lua/mason-lspconfig/server_configurations/drools_lsp/init.lua
In this way, mason-lspconfig.nvim can easily pass in the installation directory.

Second, to show the proposed code I want to sumbit to neovim/lspconfig.nvim: https://github.com/errantepiphany/nvim-lspconfig/blob/2b14816ae88d46d51260a8c86ef6184826b333e0/lua/lspconfig/server_configurations/drools_lsp.lua
You can see that it exposes a `cmd_builder` that can be reused by 3rd party installers, like mason. It's `build` function allows for an optional parameter - the `install_path`. With this proposal of mine, lspconfig.nvim has no direct ties to any installers, yet still provides sensible defaults for most people. But then it also exposes a mechanism that can be reused. I had first wanted to just re-use that code inside mason.nvim when it creates the shell_exec_wrapper, but then there would be a dependency on neovim/lspconfig.nvim from within mason.nvim, which isn't good (that's way the mason-lspconfig.nvim bridge exists).

So, what are your thoughts? Personally I think it's pretty clean, and I've successfully tested it with various configurations. My biggest challenge will be to get the neovim team to accept my PR there, since the last time I did, the reviewer was extremely picky about anything that isn't dead-simple. Which, TBH, is frustrating because there's a lot of other server configs there that are a lot longer and way more complicated.

Thank you again! :)

**Comment by williamboman** [comment]
Hey!

> So, what are your thoughts? Personally I think it's pretty clean, and I've successfully tested it with various configurations. My biggest challenge will be to get the neovim team to accept my PR there, since the last time I did, the reviewer was extremely picky about anything that isn't dead-simple. Which, TBH, is frustrating because there's a lot of other server configs there that are a lot longer and way more complicated.

I sort of have to side with the lspconfig maintainers here. In my experience most (all?) server configurations in lspconfig really only include the minimum to successfully launch a server with sensible defaults. Some servers require quite some extra client-side logic which may result in large files but I feel like that is entirely internalized. Introducing a new cmd builder API that is meant to be used by 3rd parties seems like bending the scope of lspconfig a bit too much, so I understand if they're not too happy about it (someone will have to maintain it after the fact).

I think the best approach would be to cover the general use cases only, and ignore specific edge cases. I feel like the fact that Mason provides a "convenience" executable is overcomplicating things a bit.

Wouldn't simply the following be enough?

```lua
-- mason-lspconfig/lua/server_configurations/drools-lsp/init.lua
return function ()
  return {
    drools = {
       jar = "<path_to_mason_jar_file>"
    }
  }
end
```

Should the user provide a `drools.jar` setting themselves, it will [have precedence over](https://github.com/williamboman/mason-lspconfig.nvim/blob/main/lua/mason-lspconfig/lspconfig_hook.lua#L64) the above.

**Reply by errantepiphany** [reply]
The behavior I'm seeing is opposite of what I would expect to see in the code you referenced. It appears that whatever "node" is set in `mason-lspconfig.nvim/lua/server_configurations/drools-lsp/init.lua` will _always_ override whatever matching node that the user sets in their neovim init. The tables do get merged, but the user's config settings only get respected if they're _not_ also in the default config we provide above. I'm scratching my head on this...

**Reply by williamboman** [reply]
Hmmmm that doesn't seem right 😱, I've so far relied on tests for this. Let me dig deeper

**Reply by errantepiphany** [reply]
I'm going to replace my neovim init.lua with a minimal one as a reproducer, and test again with that.

**Reply by errantepiphany** [reply]
Okay, I confirmed the behavior I described (opposite as it seems it should be) with a vanilla neovim init.lua. If you care to try, I can send that init.lua to you if you want, but then you'll have to change the plugin paths to your local development ones, after pulling in my latest changes (per the conversation thread below). I can submit PRs for those now so you have something to pull in locally. Is this too much of a pain in the butt? I don't want to cause you time and trouble, however this _might_ be something wrong with mason. (I'm not saying it is, but it might be. It could also be with unexpected behavior of `on_new_config` that I'm using in neovim/lspconfig.nvim?) I can also do a Zoom+screen share if that's easier.

**Comment by errantepiphany** [comment]
@williamboman Okay, I hear you. If they're saying it and you're saying it, it must just be me overcomplicating things, LOL. So:

1. In `neovim/lspconfig.nvim`, remove the cmd_builder, although I still need to change the config to be outside of the settings section, as you mentioned. I wish I had picked up on that sooner, but I doubt I'll be breaking it for anyone since my team is probably all who's using it so far. Besides, it's only on master and not even tagged/released yet. ([source](https://github.com/errantepiphany/nvim-lspconfig/blob/7a945cc466448a617aa3aaedb02ae5b9e836d611/lua/lspconfig/server_configurations/drools_lsp.lua))
2. In `williamboman/mason-lspconfig.nvim`, simply merge the config pointing to where the jar file was installed on behalf of the user. I can see that the java and java opts get respected if the user configures them, however if the user specified the jar location themselves, it will always get clobbered by this. I think that's okay, since they do get the benefit of zero-config. (Besides, they're using mason to install it, why not use it?) Now if they _really_ want to define a different location of the jar, then they can still specify the entire cmd themselves. ([source](https://github.com/errantepiphany/mason-lspconfig.nvim/blob/0fc926357de5edbf4f09a3358aaac299f109a7f1/lua/mason-lspconfig/server_configurations/drools_lsp/init.lua))
3. In `williamboman/mason.nvim`, remove creating the shell_exec_wrapper from the mason installer, since it isn't getting used. In fact, if we did create a script and default the cmd it, then the user wouldn't ever be able to specify anything, since the cmd we specify would always override everything else. ([source](https://github.com/errantepiphany/mason.nvim/blob/7d7c08064d044b2bb3ef991afa07d0a93df98677/lua/mason-registry/drools-lsp/init.lua))

All good now, ya' think? If so we'll have to wait to merge these until after the neovim/lspconfig.nvim maintainers merge my change there.

**Reply by williamboman** [reply]
Looks solid to me!

**Reply by errantepiphany** [reply]
Awesome. Here are the mason PRs. The can be merged independent of the neovim PR now, since I removed that `cmd_builder` coupling. Also, even though the config location has moved up from settings, that still won't effect this code, as then this config merge wouldn't even match (which is fine since this file didn't even exist in the first place). So nothing breaks with these PRs.
https://github.com/williamboman/mason.nvim/pull/925
https://github.com/williamboman/mason-lspconfig.nvim/pull/132
